sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsContainer.java

```diff
@@ -0,0 +1,201 @@
+ *
+ * <p>A thread-local variable holds the {@link MetricsContainer} that should be used for any
+ * metric updates produced from the executing code by interactions with the user-facing metric
+ * interfaces (eg., {@link Counter}).
+ *
+ * <p>For a given runner to support metrics it is currently necessary to do 3 things:
+ * <ol>
+ *   <li>Create a {@link MetricsContainer} for each step that metrics will be reported at.
+ *   </li>
+ *   <li>Make sure to call {@link #setMetricsContainer}
+ *   </li>
+ *   <li>Use {@link #getCumulative()} to get {@link org.apache.beam.sdk.metrics.MetricUpdates}
+ *   representing all the metric changes and report/aggregate those appropriately.
+ *   </li>
+ *   <li>Optionally, use {@link #getUpdates()} to get only the changes to a metric since the last
+ *   time and report those periodically. After reporting deltas and before the next call to
+ *   {@link #getUpdates()} the runner should invoke {@link #commitUpdates} to indicate that those
+ *   deltas have been incorporated.
+ *   </li>
+ * </ol>
+ *
+ * <p>This class is thread-safe.
+ */
+@Experimental(Kind.METRICS)
+public class MetricsContainer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MetricsContainer.class);
+  private static final AtomicBoolean REPORTED_MISSING_CONTAINER = new AtomicBoolean(false);
+
+  private static final ThreadLocal<MetricsContainer> CONTAINER_FOR_THREAD =
+      new ThreadLocal<MetricsContainer>();
+
+  private final String stepName;
+
+  private MetricsMap<MetricName, CounterCell> counters =
+      new MetricsMap<>(new MetricsMap.Factory<MetricName, CounterCell>() {
+        @Override
+        public CounterCell createInstance(MetricName unusedKey) {
+          return new CounterCell();
+        }
+      });
+
+  private MetricsMap<MetricName, DistributionCell> distributions =
+      new MetricsMap<>(new MetricsMap.Factory<MetricName, DistributionCell>() {
+        @Override
+        public DistributionCell createInstance(MetricName unusedKey) {
+          return new DistributionCell();
+        }
+      });
+
+  /**
+   * Create a new {@link MetricsContainer} associated with the given {@code stepName}.
+   */
+  public MetricsContainer(String stepName) {
+    this.stepName = stepName;
+  }
+
+  /**
+   * Set the {@link MetricsContainer} for the current thread.
+   */
+  public static void setMetricsContainer(MetricsContainer container) {
+    CONTAINER_FOR_THREAD.set(container);
+  }
+
+  /**
+   * Clear the {@link MetricsContainer} for the current thread.
+   */
+  public static void unsetMetricsContainer() {
+    CONTAINER_FOR_THREAD.remove();
+  }
+
+  /**
+   * Return the {@link MetricsContainer} for the current thread.
+   */
+  public static MetricsContainer getCurrentContainer() {
+    MetricsContainer container = CONTAINER_FOR_THREAD.get();
+    if (container == null && REPORTED_MISSING_CONTAINER.compareAndSet(false, true)) {
+      LOGGER.error("Unable to get {} for the current thread.\n"
```

Comment:
`setMetricsContainer` needs to be kept lean in case it is run between every element (which is possible). I'll add a static `setMetricsSupported` that a runner which supports metrics can use to report the less-scary message.